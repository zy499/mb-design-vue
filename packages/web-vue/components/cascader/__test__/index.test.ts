import { mount } from '@vue/test-utils';
import Cascader from '../cascader.vue';

const options = [
  {
    value: 'beijing',
    label: 'Beijing',
    children: [
      {
        value: 'chaoyang',
        label: 'ChaoYang',
      },
      {
        value: 'haidian',
        label: 'Haidian',
      },
    ],
  },
  {
    value: 'shanghai',
    label: 'Shanghai',
    children: [
      {
        value: 'huangpu',
        label: 'Huangpu',
      },
    ],
  },
];

describe('Cascader', () => {
  test('should render panel', async () => {
    const wrapper = mount(Cascader, {
      props: {
        options,
      },
    });
    await wrapper.find('input').trigger('click');
    const panel = wrapper.findComponent({ name: 'BaseCascaderPanel' });
    expect(panel.html()).toMatchSnapshot();
    await panel.find('.arco-cascader-option').trigger('click');
    expect(panel.html()).toMatchSnapshot();
  });

  test('should render panel (multiple)', async () => {
    const wrapper = mount(Cascader, {
      props: {
        options,
        multiple: true,
      },
    });
    await wrapper.find('input').trigger('click');
    const panel = wrapper.findComponent({ name: 'BaseCascaderPanel' });
    expect(panel.html()).toMatchSnapshot();
    await panel.find('.arco-cascader-option').trigger('click');
    expect(panel.html()).toMatchSnapshot();
  });

  test('should render search panel', async () => {
    const wrapper = mount(Cascader, {
      props: {
        options,
        allowSearch: true,
      },
    });

    const input = wrapper.find('input');
    await input.trigger('click');
    await input.setValue('a');
    const panel = wrapper.findComponent({ name: 'CascaderSearchPanel' });
    expect(panel.html()).toMatchSnapshot();
  });

  test('should emit change event', async () => {
    const wrapper = mount(Cascader, {
      props: {
        options,
      },
    });

    await wrapper.find('input').trigger('click');
    const panel = wrapper.findComponent({ name: 'BaseCascaderPanel' });
    await panel.find('.arco-cascader-option').trigger('click');
    await panel.findAll('.arco-cascader-option')[2].trigger('click');
    expect(wrapper.emitted('change')?.[0]).toEqual(['chaoyang']);
  });

  test('should emit change event (pathMode)', async () => {
    const wrapper = mount(Cascader, {
      props: {
        options,
        pathMode: true,
      },
    });

    await wrapper.find('input').trigger('click');
    const panel = wrapper.findComponent({ name: 'BaseCascaderPanel' });
    await panel.find('.arco-cascader-option').trigger('click');
    await panel.findAll('.arco-cascader-option')[2].trigger('click');
    expect(wrapper.emitted('change')?.[0]).toEqual([['beijing', 'chaoyang']]);
  });

  test('should emit change event (multiple)', async () => {
    const wrapper = mount(Cascader, {
      props: {
        options,
        multiple: true,
      },
    });

    await wrapper.find('input').trigger('click');
    const panel = wrapper.findComponent({ name: 'BaseCascaderPanel' });
    await panel.find('.arco-cascader-option').trigger('click');
    await panel.findAllComponents({ name: 'Checkbox' })[2].trigger('click');
    expect(wrapper.emitted('change')?.[0]).toEqual([['chaoyang']]);
  });

  test('should support keyboard action', async () => {
    const wrapper = mount(Cascader, {
      props: {
        options,
      },
    });

    const input = wrapper.find('input');
    await input.trigger('click');
    const dropdown = wrapper.findComponent({ name: 'BaseCascaderPanel' });

    await input.trigger('keydown', { key: 'ArrowDown' });
    expect(dropdown.find('.arco-cascader-option-active').text()).toBe(
      'Beijing'
    );

    await input.trigger('keydown', { key: 'ArrowRight' });
    expect(dropdown.findAll('.arco-cascader-panel-column')).toHaveLength(2);
    expect(dropdown.findAll('.arco-cascader-option-active')[1].text()).toBe(
      'ChaoYang'
    );

    await input.trigger('keydown', { key: 'ArrowDown' });
    expect(dropdown.findAll('.arco-cascader-option-active')[1].text()).toBe(
      'Haidian'
    );

    await input.trigger('keydown', { key: 'Enter' });
    expect(wrapper.emitted('change')?.[0]).toEqual(['haidian']);
  });
});
