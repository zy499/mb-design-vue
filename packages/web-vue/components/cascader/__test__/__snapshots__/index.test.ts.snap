// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cascader should render panel (multiple) 1`] = `
"<transition-group-stub>
  <div class=\\"mb-cascader-panel-column\\" style=\\"z-index: 2;\\">
    <ul role=\\"menu\\" class=\\"mb-cascader-list mb-cascader-list-multiple\\">
      <li tabindex=\\"0\\" role=\\"menuitem\\" aria-disabled=\\"false\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\" title=\\"Beijing\\" class=\\"mb-cascader-option\\"><label class=\\"mb-checkbox\\"><input type=\\"checkbox\\" class=\\"mb-checkbox-target\\"><span class=\\"mb-icon-hover mb-checkbox-icon-hover\\"><div class=\\"mb-checkbox-icon\\"><!----></div></span>
          <!---->
        </label>
        <!---->
        <div class=\\"mb-cascader-option-label\\">Beijing<svg viewBox=\\"0 0 48 48\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" stroke=\\"currentColor\\" class=\\"mb-icon mb-icon-right\\" stroke-width=\\"4\\" stroke-linecap=\\"butt\\" stroke-linejoin=\\"miter\\">
            <path d=\\"m16 39.513 15.556-15.557L16 8.4\\"></path>
          </svg></div>
      </li>
      <li tabindex=\\"0\\" role=\\"menuitem\\" aria-disabled=\\"false\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\" title=\\"Shanghai\\" class=\\"mb-cascader-option\\"><label class=\\"mb-checkbox\\"><input type=\\"checkbox\\" class=\\"mb-checkbox-target\\"><span class=\\"mb-icon-hover mb-checkbox-icon-hover\\"><div class=\\"mb-checkbox-icon\\"><!----></div></span>
          <!---->
        </label>
        <!---->
        <div class=\\"mb-cascader-option-label\\">Shanghai<svg viewBox=\\"0 0 48 48\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" stroke=\\"currentColor\\" class=\\"mb-icon mb-icon-right\\" stroke-width=\\"4\\" stroke-linecap=\\"butt\\" stroke-linejoin=\\"miter\\">
            <path d=\\"m16 39.513 15.556-15.557L16 8.4\\"></path>
          </svg></div>
      </li>
    </ul>
  </div>
</transition-group-stub>"
`;

exports[`Cascader should render panel (multiple) 2`] = `
"<transition-group-stub>
  <div class=\\"arco-cascader-panel-column\\" style=\\"z-index: 2;\\">
    <ul role=\\"menu\\" class=\\"arco-cascader-list arco-cascader-list-multiple\\">
      <li tabindex=\\"0\\" role=\\"menuitem\\" aria-disabled=\\"false\\" aria-haspopup=\\"true\\" aria-expanded=\\"true\\" title=\\"Beijing\\" class=\\"arco-cascader-option arco-cascader-option-active\\"><label class=\\"arco-checkbox\\"><input type=\\"checkbox\\" class=\\"arco-checkbox-target\\"><span class=\\"arco-icon-hover arco-checkbox-icon-hover\\"><div class=\\"arco-checkbox-icon\\"><!----></div></span>
          <!---->
        </label>
        <!---->
        <div class=\\"arco-cascader-option-label\\">Beijing<svg viewBox=\\"0 0 48 48\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" stroke=\\"currentColor\\" class=\\"arco-icon arco-icon-right\\" stroke-width=\\"4\\" stroke-linecap=\\"butt\\" stroke-linejoin=\\"miter\\">
            <path d=\\"m16 39.513 15.556-15.557L16 8.4\\"></path>
          </svg></div>
      </li>
      <li tabindex=\\"0\\" role=\\"menuitem\\" aria-disabled=\\"false\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\" title=\\"Shanghai\\" class=\\"arco-cascader-option\\"><label class=\\"arco-checkbox\\"><input type=\\"checkbox\\" class=\\"arco-checkbox-target\\"><span class=\\"arco-icon-hover arco-checkbox-icon-hover\\"><div class=\\"arco-checkbox-icon\\"><!----></div></span>
          <!---->
        </label>
        <!---->
        <div class=\\"arco-cascader-option-label\\">Shanghai<svg viewBox=\\"0 0 48 48\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" stroke=\\"currentColor\\" class=\\"arco-icon arco-icon-right\\" stroke-width=\\"4\\" stroke-linecap=\\"butt\\" stroke-linejoin=\\"miter\\">
            <path d=\\"m16 39.513 15.556-15.557L16 8.4\\"></path>
          </svg></div>
      </li>
    </ul>
  </div>
  <div class=\\"arco-cascader-panel-column\\" style=\\"z-index: 1;\\">
    <ul role=\\"menu\\" class=\\"arco-cascader-list arco-cascader-list-multiple\\">
      <li tabindex=\\"0\\" role=\\"menuitem\\" aria-disabled=\\"false\\" aria-haspopup=\\"false\\" aria-expanded=\\"false\\" title=\\"ChaoYang\\" class=\\"arco-cascader-option\\"><label class=\\"arco-checkbox\\"><input type=\\"checkbox\\" class=\\"arco-checkbox-target\\"><span class=\\"arco-icon-hover arco-checkbox-icon-hover\\"><div class=\\"arco-checkbox-icon\\"><!----></div></span>
          <!---->
        </label>
        <!---->
        <div class=\\"arco-cascader-option-label\\">ChaoYang
          <!---->
        </div>
      </li>
      <li tabindex=\\"0\\" role=\\"menuitem\\" aria-disabled=\\"false\\" aria-haspopup=\\"false\\" aria-expanded=\\"false\\" title=\\"Haidian\\" class=\\"arco-cascader-option\\"><label class=\\"arco-checkbox\\"><input type=\\"checkbox\\" class=\\"arco-checkbox-target\\"><span class=\\"arco-icon-hover arco-checkbox-icon-hover\\"><div class=\\"arco-checkbox-icon\\"><!----></div></span>
          <!---->
        </label>
        <!---->
        <div class=\\"arco-cascader-option-label\\">Haidian
          <!---->
        </div>
      </li>
    </ul>
  </div>
</transition-group-stub>"
`;

exports[`Cascader should render panel 1`] = `
"<transition-group-stub>
  <div class=\\"mb-cascader-panel-column\\" style=\\"z-index: 2;\\">
    <ul role=\\"menu\\" class=\\"mb-cascader-list\\">
      <li tabindex=\\"0\\" role=\\"menuitem\\" aria-disabled=\\"false\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\" title=\\"Beijing\\" class=\\"mb-cascader-option\\">
        <!---->
        <!---->
        <div class=\\"mb-cascader-option-label\\">Beijing<svg viewBox=\\"0 0 48 48\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" stroke=\\"currentColor\\" class=\\"mb-icon mb-icon-right\\" stroke-width=\\"4\\" stroke-linecap=\\"butt\\" stroke-linejoin=\\"miter\\">
            <path d=\\"m16 39.513 15.556-15.557L16 8.4\\"></path>
          </svg></div>
      </li>
      <li tabindex=\\"0\\" role=\\"menuitem\\" aria-disabled=\\"false\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\" title=\\"Shanghai\\" class=\\"mb-cascader-option\\">
        <!---->
        <!---->
        <div class=\\"mb-cascader-option-label\\">Shanghai<svg viewBox=\\"0 0 48 48\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" stroke=\\"currentColor\\" class=\\"mb-icon mb-icon-right\\" stroke-width=\\"4\\" stroke-linecap=\\"butt\\" stroke-linejoin=\\"miter\\">
            <path d=\\"m16 39.513 15.556-15.557L16 8.4\\"></path>
          </svg></div>
      </li>
    </ul>
  </div>
</transition-group-stub>"
`;

exports[`Cascader should render panel 2`] = `
"<transition-group-stub>
  <div class=\\"arco-cascader-panel-column\\" style=\\"z-index: 2;\\">
    <ul role=\\"menu\\" class=\\"arco-cascader-list\\">
      <li tabindex=\\"0\\" role=\\"menuitem\\" aria-disabled=\\"false\\" aria-haspopup=\\"true\\" aria-expanded=\\"true\\" title=\\"Beijing\\" class=\\"arco-cascader-option arco-cascader-option-active\\">
        <!---->
        <!---->
        <div class=\\"arco-cascader-option-label\\">Beijing<svg viewBox=\\"0 0 48 48\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" stroke=\\"currentColor\\" class=\\"arco-icon arco-icon-right\\" stroke-width=\\"4\\" stroke-linecap=\\"butt\\" stroke-linejoin=\\"miter\\">
            <path d=\\"m16 39.513 15.556-15.557L16 8.4\\"></path>
          </svg></div>
      </li>
      <li tabindex=\\"0\\" role=\\"menuitem\\" aria-disabled=\\"false\\" aria-haspopup=\\"true\\" aria-expanded=\\"false\\" title=\\"Shanghai\\" class=\\"arco-cascader-option\\">
        <!---->
        <!---->
        <div class=\\"arco-cascader-option-label\\">Shanghai<svg viewBox=\\"0 0 48 48\\" fill=\\"none\\" xmlns=\\"http://www.w3.org/2000/svg\\" stroke=\\"currentColor\\" class=\\"arco-icon arco-icon-right\\" stroke-width=\\"4\\" stroke-linecap=\\"butt\\" stroke-linejoin=\\"miter\\">
            <path d=\\"m16 39.513 15.556-15.557L16 8.4\\"></path>
          </svg></div>
      </li>
    </ul>
  </div>
  <div class=\\"arco-cascader-panel-column\\" style=\\"z-index: 1;\\">
    <ul role=\\"menu\\" class=\\"arco-cascader-list\\">
      <li tabindex=\\"0\\" role=\\"menuitem\\" aria-disabled=\\"false\\" aria-haspopup=\\"false\\" aria-expanded=\\"false\\" title=\\"ChaoYang\\" class=\\"arco-cascader-option\\">
        <!---->
        <!---->
        <div class=\\"arco-cascader-option-label\\">ChaoYang
          <!---->
        </div>
      </li>
      <li tabindex=\\"0\\" role=\\"menuitem\\" aria-disabled=\\"false\\" aria-haspopup=\\"false\\" aria-expanded=\\"false\\" title=\\"Haidian\\" class=\\"arco-cascader-option\\">
        <!---->
        <!---->
        <div class=\\"arco-cascader-option-label\\">Haidian
          <!---->
        </div>
      </li>
    </ul>
  </div>
</transition-group-stub>"
`;

exports[`Cascader should render search panel 1`] = `
"<div class=\\"mb-cascader-panel mb-cascader-search-panel\\">
  <ul role=\\"menu\\" class=\\"mb-cascader-list mb-cascader-search-list\\">
    <li tabindex=\\"0\\" role=\\"menuitem\\" aria-disabled=\\"false\\" aria-haspopup=\\"false\\" aria-expanded=\\"false\\" title=\\"ChaoYang\\" class=\\"mb-cascader-option mb-cascader-search-option\\">
      <!---->
      <!---->
      <div class=\\"mb-cascader-option-label\\">Beijing / ChaoYang
        <!---->
      </div>
    </li>
    <li tabindex=\\"0\\" role=\\"menuitem\\" aria-disabled=\\"false\\" aria-haspopup=\\"false\\" aria-expanded=\\"false\\" title=\\"Haidian\\" class=\\"mb-cascader-option mb-cascader-search-option\\">
      <!---->
      <!---->
      <div class=\\"mb-cascader-option-label\\">Beijing / Haidian
        <!---->
      </div>
    </li>
    <li tabindex=\\"0\\" role=\\"menuitem\\" aria-disabled=\\"false\\" aria-haspopup=\\"false\\" aria-expanded=\\"false\\" title=\\"Huangpu\\" class=\\"mb-cascader-option mb-cascader-search-option\\">
      <!---->
      <!---->
      <div class=\\"mb-cascader-option-label\\">Shanghai / Huangpu
        <!---->
      </div>
    </li>
  </ul>
</div>"
`;
